datafusion/physical-plan/src/yield_stream.rs

```diff
@@ -0,0 +1,271 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::any::Any;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use crate::execution_plan::CardinalityEffect::{self, Equal};
+use crate::{
+    DisplayAs, DisplayFormatType, ExecutionPlan, PlanProperties, RecordBatchStream,
+    SendableRecordBatchStream,
+};
+use arrow::record_batch::RecordBatch;
+use arrow_schema::Schema;
+use datafusion_common::{internal_err, Result, Statistics};
+use datafusion_execution::TaskContext;
+
+use futures::{Stream, StreamExt};
+
+/// An identity stream that passes batches through as is, but yields control
+/// back to the runtime every `period` batches. This stream is useful to
+/// construct a mechanism that allows operators that do not directly cooperate
+/// with the runtime to  check/support cancellation.
+pub struct YieldStream {
+    inner: SendableRecordBatchStream,
+    batches_processed: usize,
+    period: usize,
+}
+
+impl YieldStream {
+    pub fn new(inner: SendableRecordBatchStream, mut period: usize) -> Self {
+        if period == 0 {
+            period = usize::MAX;
+        }
+        Self {
+            inner,
+            batches_processed: 0,
+            period,
+        }
+    }
+}
+
+impl Stream for YieldStream {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        if self.batches_processed >= self.period {
+            self.batches_processed = 0;
+            cx.waker().wake_by_ref();
+            return Poll::Pending;
+        }
+
+        let value = self.inner.poll_next_unpin(cx);
+        match value {
+            Poll::Ready(Some(Ok(_))) => {
+                self.batches_processed += 1;
+            }
+            Poll::Pending => {
+                self.batches_processed = 0;
+            }
+            _ => {}
+        }
+        value
+    }
+}
+
+impl RecordBatchStream for YieldStream {
+    fn schema(&self) -> Arc<Schema> {
+        self.inner.schema()
+    }
+}
+
+/// This operator wraps any other execution plan and to "adapt" it to cooperate
+/// with the runtime by yielding control back to the runtime every `frequency`
+/// batches. This is useful for operators that do not natively support yielding
+/// control, allowing them to be used in a runtime that requires yielding for
+/// cancellation or other purposes.
```

Comment:
I think it would also help to explain here how to avoid this Exec in your plan

```suggestion
/// cancellation or other purposes.
///
/// # Note
/// If your ExecutionPlan periodically yields control back to the scheduler
/// implement [`ExecutionPlan::with_cooperative_yields`] to avoid the need for this
/// node.
```